Package requires Python >= 3.8

As highlighted in #2 the option `dirs_exist_ok` in the `shutil.copytree` function is used, which was introduced in Python 3.8, this means this package is currently only supporting Python 3.8 and higher.

With #4 we now see the incompatibility in the Pylint output:

```
=================================== FAILURES ===================================
_______________ [pylint] src/simplegallery_bulkcreation/core.py ________________
E: 97, 4: Unexpected keyword argument 'dirs_exist_ok' in function call (unexpected-keyword-arg)
___________ [pylint] src/simplegallery_bulkcreation/galleryclass.py ____________
E:112, 8: Unexpected keyword argument 'dirs_exist_ok' in function call (unexpected-keyword-arg)
E:117, 8: Unexpected keyword argument 'dirs_exist_ok' in function call (unexpected-keyword-arg)

--- a/.github/workflows/python-test.yml
+++ b/.github/workflows/python-test.yml
@@ -18,15 +18,12 @@ jobs:
       fail-fast: false
       matrix:
         os: [ubuntu-latest, macos-latest, windows-latest]
-        python-version: [3.8, 3.9]
+        python-version: [3.6, 3.7, 3.8, 3.9]
         experimental: [false]
-        include:
-          - python-version: 3.6
-            os: ubuntu-latest
-            experimental: true
-          - python-version: 3.7
-            os: ubuntu-latest
-            experimental: true
+        # include:
+        #   - python-version: 3.10
+        #     os: ubuntu-latest
+        #     experimental: true
 
     steps:
     - uses: actions/checkout@v2

--- a/.github/workflows/python-test.yml
+++ b/.github/workflows/python-test.yml
@@ -29,7 +29,7 @@ jobs:
         python -m pip install -U pip
         python -m pip install -U setuptools wheel
         python -m pip install -r requirements.txt
-    - name: Test with black, flake8, pylint & pytest
-      run: pytest -v --black --flake8 --pylint --pylint-error-types=EF --cov=src/simplegallery_bulkcreation
+    - name: Running tests
+      run: pytest
     - name: Run gallery-bulkcreation command
       run: gallery-bulkcreation example/config-example.ini

--- a/.github/workflows/python-test.yml
+++ b/.github/workflows/python-test.yml
@@ -5,7 +5,7 @@ name: unittests
 
 on:
   push:
-    branches: [ master ]
+    branches: [ "*" ]
   pull_request:
     branches: [ master ]
 
@@ -13,10 +13,20 @@ jobs:
   build:
 
     runs-on: ${{ matrix.os }}
+    continue-on-error: ${{ matrix.experimental }}
     strategy:
+      fail-fast: false
       matrix:
         os: [ubuntu-latest, macos-latest, windows-latest]
         python-version: [3.8, 3.9]
+        experimental: [false]
+        include:
+          - python-version: 3.6
+            os: ubuntu-latest
+            experimental: true
+          - python-version: 3.7
+            os: ubuntu-latest
+            experimental: true
 
     steps:
     - uses: actions/checkout@v2

--- a/.github/workflows/python-test.yml
+++ b/.github/workflows/python-test.yml
@@ -41,5 +41,5 @@ jobs:
         python -m pip install -r requirements.txt
     - name: Running tests
       run: pytest
-    - name: Run gallery-bulkcreation command
-      run: gallery-bulkcreation example/config-example.ini
+    - name: Run gallery-bulkcreation command twice
+      run: gallery-bulkcreation example/config-example.ini; gallery-bulkcreation example/config-example.ini

--- a/.github/workflows/python-test.yml
+++ b/.github/workflows/python-test.yml
@@ -1,7 +1,7 @@
 # This workflow will install Python dependencies, run tests and lint with a variety of Python versions
 # For more information see: https://help.github.com/actions/language-and-framework-guides/using-python-with-github-actions
 
-name: unittests
+name: build
 
 on:
   push:

--- a/README.rst
+++ b/README.rst
@@ -1,8 +1,8 @@
-.. image:: https://github.com/hille721/simple-photo-gallery-bulkcreation/workflows/build/badge.svg
-    :alt: Built Status
-    :target: https://github.com/hille721/simple-photo-gallery-bulkcreation/actions?query=workflow%3Abuild
+.. image:: https://github.com/hille721/simple-photo-gallery-bulkcreation/actions/workflows/python-test.yml/badge.svg
+    :alt: build
+    :target: https://github.com/hille721/simple-photo-gallery-bulkcreation/actions/workflows/python-test.yml
 .. image:: https://img.shields.io/pypi/v/simple-photo-gallery-bulkcreation.svg
-    :alt: PyPI-Server
+    :alt: pypi
     :target: https://pypi.org/project/simple-photo-gallery-bulkcreation/
 
 =================================
@@ -64,14 +64,13 @@ After the creation of a `config.ini`, the creation is pretty easy via running fo
 
     gallery-bulkcreation config-example.ini
 
-If everything works correctly you can preview the result via following command:
+If everything works correctly you can preview the result, e.g. via running a simple Python server:
 
 .. code-block::
 
-    python3 -m http.server --directory $GALLERY_ROOT/public
+    python3 -m http.server --directory GALLERY_ROOT/public
 
-and then checking it under the url http://localhost:8000 in your browser.
-
+and then checking it in your browser under http://localhost:8000.
 
 If you add photos to already existing galleries or add new galleries in the `config.ini` you can simple update the gallery by running the command again.
 

--- a/setup.cfg
+++ b/setup.cfg
@@ -10,9 +10,9 @@ long-description-content-type = text/x-rst; charset=UTF-8
 url = https://github.com/hille721/simple-photo-gallery-bulkcreation/
 project-urls =
     Documentation = https://github.com/hille721/simple-photo-gallery-bulkcreation/
-platforms = Linux
+platforms = any
 classifiers =
-    Development Status :: 4 - Beta
+    Development Status :: 5 - Production/Stable
     Programming Language :: Python
     Programming Language :: Python :: 3
     Programming Language :: Python :: 3.6
@@ -20,6 +20,8 @@ classifiers =
     Programming Language :: Python :: 3.8
     Programming Language :: Python :: 3.9
     Operating System :: POSIX :: Linux
+    Operating System :: MacOS
+    Operating System :: Microsoft :: Windows
     License :: OSI Approved :: MIT License
 
 [options]
@@ -62,3 +64,7 @@ exclude =
     docs/conf.py
     tests/*
 max-line-length = 99
+
+[tool:pytest]
+addopts = -v --black --flake8 --pylint --pylint-error-types=EF --cov=src/simplegallery_bulkcreation
+

--- a/src/simplegallery_bulkcreation/core.py
+++ b/src/simplegallery_bulkcreation/core.py
@@ -10,7 +10,11 @@ import sys
 import jinja2
 import pkg_resources
 
+from simplegallery_bulkcreation.utils import copytree
 from simplegallery_bulkcreation.galleryclass import SimpleGallery, spg_common
+
+if sys.version_info.major < 3:
+    sys.exit("Python 2 not supported!")
 
 
 def parse_args():
@@ -96,7 +100,7 @@ def create_overview_public(root_dir, data_path, defaults, galleries):
     spg_common.log(f"Creating overview index page under {public_dir}...")
 
     # move css into public directory
-    shutil.copytree(data_dir / "public", public_dir, dirs_exist_ok=True)
+    copytree(data_dir / "public", public_dir)
     if title_photo.exists():
         shutil.copy(title_photo, public_dir / title_photo.name)
         defaults["title_photo"] = title_photo.name

--- a/src/simplegallery_bulkcreation/core.py
+++ b/src/simplegallery_bulkcreation/core.py
@@ -10,7 +10,7 @@ import sys
 import jinja2
 import pkg_resources
 
-from simplegallery_bulkcreation.galleryclass import SimpleGallery
+from simplegallery_bulkcreation.galleryclass import SimpleGallery, spg_common
 
 
 def parse_args():
@@ -93,6 +93,8 @@ def create_overview_public(root_dir, data_path, defaults, galleries):
     data_dir = Path(data_path)
     title_photo = Path(defaults["title_photo"])
 
+    spg_common.log(f"Creating overview index page under {public_dir}...")
+
     # move css into public directory
     shutil.copytree(data_dir / "public", public_dir, dirs_exist_ok=True)
     if title_photo.exists():
@@ -101,6 +103,7 @@ def create_overview_public(root_dir, data_path, defaults, galleries):
 
     # generate root index.html
     write_overview_index(public_dir, defaults, galleries, data_dir / "templates")
+    spg_common.log(f"Overview index page {public_dir}/index.html created successfully!")
 
 
 def main():
@@ -120,14 +123,20 @@ def main():
 
     defaults, galleries = read_config(config_path=config_path)
     root_dir = defaults["gallery_root"]
+    seperator_line = "----------------------------------------"
 
+    spg_common.log(seperator_line)
     create_overview_public(root_dir, data_path, defaults, galleries)
+    spg_common.log(seperator_line)
 
     for gallery_conf in galleries:
         # TODO: Exception handling!
         try:
             gallery = SimpleGallery(root_dir, **gallery_conf)
             gallery.gallery_init()
+            spg_common.log(seperator_line)
             gallery.gallery_build()
+            spg_common.log(seperator_line)
         except Exception as exc:
             sys.exit(str(exc))
+    spg_common.log(f"All galleries were built successfully. Open {root_dir}/public.")

--- a/src/simplegallery_bulkcreation/galleryclass.py
+++ b/src/simplegallery_bulkcreation/galleryclass.py
@@ -10,6 +10,8 @@ import simplegallery.common as spg_common
 from simplegallery import gallery_build
 from simplegallery.logic.gallery_logic import get_gallery_logic
 
+from simplegallery_bulkcreation.utils import copytree
+
 
 def _pltostr(path):
     """
@@ -111,15 +113,13 @@ class SimpleGallery:
         """
         # Copy the public and templates folder
         spg_common.log("Copying gallery template files...")
-        shutil.copytree(
+        copytree(
             Path(pkg_resources.resource_filename("simplegallery", "data/templates")),
             self.gallery_path / "templates",
-            dirs_exist_ok=True,
         )
-        shutil.copytree(
+        copytree(
             Path(pkg_resources.resource_filename("simplegallery", "data/public")),
             self.public_gallery_dir,
-            dirs_exist_ok=True,
         )
 
         photos_dir = self.public_gallery_dir / "images" / "photos"

--- a/src/simplegallery_bulkcreation/galleryclass.py
+++ b/src/simplegallery_bulkcreation/galleryclass.py
@@ -94,7 +94,9 @@ class SimpleGallery:
             json.dumps(gallery_config, indent=4, separators=(",", ": "))
         )
 
-        spg_common.log("Gallery config stored in gallery.json")
+        spg_common.log(
+            f"Gallery config for {self.name} stored in {self.gallery_config_path}"
+        )
 
         return True
 
@@ -196,8 +198,6 @@ class SimpleGallery:
         spg_common.log(f"Creating the index.html for {self.name}...")
         gallery_build.build_html(gallery_config)
 
-        spg_common.log(
-            f"The gallery {self.name} was built successfully. Open {self.public_gallery_dir}"
-        )
+        spg_common.log(f"The gallery {self.name} was built successfully.")
 
         return True

--- a/src/simplegallery_bulkcreation/utils.py
+++ b/src/simplegallery_bulkcreation/utils.py
@@ -1,0 +1,34 @@
+""" some general helping functions """
+import sys
+
+if sys.version_info.major < 3:
+    sys.exit("Python 2 not supported!")
+if sys.version_info.major == 3 and sys.version_info.minor < 8:
+    from distutils.dir_util import copy_tree
+
+    LEGACY_SUPPORT = True
+else:
+    import shutil
+
+    LEGACY_SUPPORT = False
+
+
+def copytree(src, dst):
+    """
+    Wrapper around `shutil.copytree` function to support also Python < 3.8 (see #3).
+
+    Backround:
+        The `shutil.copytree` option `dirs_exist_ok=True` was introduced in Python 3.8
+        To support Python 3.6 and Python 3.7 as well, we have to do a workaround
+        and using the `copy_tree` function of the `distutils.dir_util` module.
+
+    Args:
+        src (Path): source path
+        dst (Path): destination path
+    """
+    if LEGACY_SUPPORT:
+        return copy_tree(str(src.absolute()), str(dst.absolute()))
+
+    return shutil.copytree(  # pylint: disable=unexpected-keyword-arg
+        src, dst, dirs_exist_ok=True
+    )

